### PR TITLE
fix: deadlock when closing transient push query

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 
fixes: https://github.com/confluentinc/ksql/issues/4296

The produce side not calls `offer` in a loop, with a short timeout, to try and put the row into the blocking queue. When the consume side closes the query, e.g. on an `EOFException` if the user has closed the connection, the query first closes the queue; setting a flag the producers are checking on each loop; causing any producers to exit the loop. Then it can safely close the KS topology.

### Testing done 

Added unit tests and manual testing

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

**Comment by big-andy-coates** [comment]
We may want to backport this onto 5.4.1. But I'll do that manually.

File: ksql-engine/src/main/java/io/confluent/ksql/query/BlockingQueryQueue.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import org.apache.kafka.streams.KeyValue;
+
+/**
+ * The queue between the Kafka-streams topology and the client connection.
+ *
+ * <p>The KS topology writes to the queue from its {@code StreamThread}, while the KSQL server
+ * thread that is servicing the client request reads from the queue and writes to the client
+ * socket.
+ */
+public interface BlockingQueryQueue {
```
Comment (agavra) [inline ksql-engine/src/main/java/io/confluent/ksql/query/BlockingQueryQueue.java]: nit: the name makes it sound like a it's a queue of queries, not a blocking queue for a query. thoughts on `BlockingEventQueue` or `BlockingRowQueue` (a queue that fills up with events/rows)?

File: ksql-engine/src/main/java/io/confluent/ksql/query/BlockingQueryQueue.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+  void setLimitHandler(LimitHandler limitHandler);
```
Comment (agavra) [inline ksql-engine/src/main/java/io/confluent/ksql/query/BlockingQueryQueue.java]: would be nice to also document this (or add documentation to `LimitHandler`) to describe when exactly the limit is considered reached

File: ksql-engine/src/main/java/io/confluent/ksql/query/BlockingQueryQueue.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+   *
+   * @see BlockingQueryQueue#poll(long, TimeUnit)
+   */
+  KeyValue<String, GenericRow> poll(long timeout, TimeUnit unit)
+      throws InterruptedException;
+
+  /**
+   * Drain the queue to the supplied {@code collection}.
+   *
+   * @see BlockingQueryQueue#drainTo(Collection)
```
Comment (agavra) [inline ksql-engine/src/main/java/io/confluent/ksql/query/BlockingQueryQueue.java]: self-referential javadoc?

File: ksql-engine/src/main/java/io/confluent/ksql/query/BlockingQueryQueue.java (hunk 1 above)
Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/query/BlockingQueryQueue.java]: sure.

File: ksql-engine/src/main/java/io/confluent/ksql/query/BlockingQueryQueue.java (hunk 2 above)
Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/query/BlockingQueryQueue.java]: sure

File: ksql-engine/src/main/java/io/confluent/ksql/query/BlockingQueryQueue.java (hunk 3 above)
Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/query/BlockingQueryQueue.java]: oops

**Pull request merged**